### PR TITLE
Add support for XEP-0359: Stable and Unique Stanza IDs

### PR DESCRIPTION
**Pull request by vanitasvitae**
Greetings from Brussels and the XSF-Summit!

TODO:
* [x] Fix breaking of MAMIntegrationTest

### CODE REVIEW

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+        return disco.includesFeature(NAMESPACE);
+    }
+
+    /**
+     * Create an origin-id element with a random UUID.
+     *
+     * @return origin-id element.
+     */
+    public static OriginIdElement createOriginId() {
+        return new OriginIdElement(UUID.randomUUID().toString());
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Should be simply the behavior of `OriginIdElement`'s non-argument constructor.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Add an origin-id element to a stanza and set the stanzas id to the same id as in the origin-id element.
+     *
+     * @param stanza stanza.
+     */
+    public static void addOriginId(Stanza stanza) {
+        OriginIdElement originId = createOriginId();
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Should be part of `OriginIdElement` API, not the manger. As a rule of thumb: Static methods in managers dealing soley with elements usually belong to those elements.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+        stanza.setStanzaId(originId.getId());
+    }
+
+    /**
+     * Return true, if a stanza contains a stanza-id element.
+     *
+     * @param stanza stanza
+     * @return true if message contains stanza-id element, otherwise false.
+     */
+    public static boolean hasStanzaId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Move to `StanzaIdElement`

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Return the stanza-id element of a stanza.
+     *
+     * @param stanza stanza
+     * @return stanza-id element of a jid, or null if absent.
+     */
+    public static StanzaIdElement getStanzaId(Stanza stanza) {
+        return stanza.getExtension(StanzaIdElement.ELEMENT, NAMESPACE);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Move to StanzaIdElement

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Return true, if the stanza contains a origin-id element.
+     *
+     * @param stanza stanza
+     * @return true if the stanza contains a origin-id, false otherwise.
+     */
+    public static boolean hasOriginId(Stanza stanza) {
+        return getOriginId(stanza) != null;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Move to OrigindElement

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+    }
+
+    /**
+     * Return the origin-id element of a stanza or null, if absent.
+     *
+     * @param stanza stanza
+     * @return origin-id element
+     */
+    public static OriginIdElement getOriginId(Stanza stanza) {
+        return stanza.getExtension(OriginIdElement.ELEMENT, NAMESPACE);
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Move to OrigindElement

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jivesoftware.smackx.sid.element;
+
+import org.jivesoftware.smack.packet.ExtensionElement;
+import org.jivesoftware.smack.util.XmlStringBuilder;
+import org.jivesoftware.smackx.sid.StableUniqueStanzaIdManager;
+
+public class OriginIdElement implements ExtensionElement {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java]: There should be a common superclass of `OriginIdElement` and `StanzaIdElement`, e.g. `UniqueAndStableIdElement`.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    public static final String ELEMENT = "stanza-id";
+    public static final String ATTR_ID = "id";
+    public static final String ATTR_BY = "by";
+
+    private final String id;
+    private final String by;
+
+    public StanzaIdElement(String id, String by) {
+        this.id = id;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java]: Missing `StringUtils.isNotNullOrEmpty()` check

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+    public static final String ELEMENT = "origin-id";
+    public static final String ATTR_ID = "id";
+
+    private final String id;
+
+    public OriginIdElement(String id) {
+        this.id = id;
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java]: Missing `StringUtils.isNotNullOrEmpty()` check

File: smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.providers [hunk 10]
```diff
@@ -211,6 +211,19 @@
...
+        <namespace>urn:xmpp:sid:0</namespace>
+        <className>org.jivesoftware.smackx.sid.provider.StanzaIdProvider</className>
+    </extensionProvider>
+    <extensionProvider>
+        <elementName>origin-id</elementName>
+        <namespace>urn:xmpp:sid:0</namespace>
+        <className>org.jivesoftware.smackx.sid.provider.OriginIdProvider</className>
+    </extensionProvider>
+
+
```
Comment (Flowdalic) [inline smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.providers]: Nit: Superfluous newline

File: smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml [hunk 11]
```diff
@@ -5,6 +5,7 @@
         <className>org.jivesoftware.smackx.iot.discovery.IoTDiscoveryManager</className>
         <className>org.jivesoftware.smackx.iot.provisioning.IoTProvisioningManager</className>
         <className>org.jivesoftware.smackx.httpfileupload.HttpFileUploadManager</className>
-		<className>org.jivesoftware.smackx.eme.ExplicitMessageEncryptionManager</className>
+        <className>org.jivesoftware.smackx.eme.ExplicitMessageEncryptionManager</className>
```
Comment (Flowdalic) [inline smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml]: Nit: Unrelated whitespace fix

**Comment by Flowdalic** [comment]
The extensions/index.md entry is also missing.

File: smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml (hunk 11 above)
Comment (vanitasvitae) [reply smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml]: I cannot fix this because my IDE doesn't use tabs.

**Comment by vanitasvitae** [comment]
> The extensions/index.md entry is also missing.
Added.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+    public OriginIdElement(String id) {
+        super(id);
+    }
+
+    /**
+     * Add an origin-id element to a stanza and set the stanzas id to the same id as in the origin-id element.
+     *
+     * @param stanza stanza.
+     */
+    public static void addOriginId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java]: This should be `Message` as the extensions is only specified for messages.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java [hunk 13]
```diff
@@ -0,0 +1,82 @@
...
+        stanza.setStanzaId(originId.getId());
+    }
+
+    /**
+     * Return true, if the stanza contains a origin-id element.
+     *
+     * @param stanza stanza
+     * @return true if the stanza contains a origin-id, false otherwise.
+     */
+    public static boolean hasOriginId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java]: This should be Message as the extensions is only specified for messages.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        return getOriginId(stanza) != null;
+    }
+
+    /**
+     * Return the origin-id element of a stanza or null, if absent.
+     *
+     * @param stanza stanza
+     * @return origin-id element
+     */
+    public static OriginIdElement getOriginId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/OriginIdElement.java]: This should be Message as the extensions is only specified for messages.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+        this.by = by;
+    }
+
+    /**
+     * Return true, if a stanza contains a stanza-id element.
+     *
+     * @param stanza stanza
+     * @return true if message contains stanza-id element, otherwise false.
+     */
+    public static boolean hasStanzaId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java]: This should be Message as the extensions is only specified for messages.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+        return getStanzaId(stanza) != null;
+    }
+
+    /**
+     * Return the stanza-id element of a stanza.
+     *
+     * @param stanza stanza
+     * @return stanza-id element of a jid, or null if absent.
+     */
+    public static StanzaIdElement getStanzaId(Stanza stanza) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/element/StanzaIdElement.java]: This should be Message as the extensions is only specified for messages.

File: smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml (hunk 11 above)
Comment (Flowdalic) [reply smack-experimental/src/main/resources/org.jivesoftware.smack.experimental/experimental.xml]: But surely you can force your IDE to produce tabs?

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+            manager = new StableUniqueStanzaIdManager(connection);
+            INSTANCES.put(connection, manager);
+        }
+        return manager;
+    }
+
+    /**
+     * Start appending origin-id elements to outgoing stanzas and add the feature to disco.
+     */
+    public void enable() {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Please make enable() and disable() synchronized and remove the shortcuts.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 18]
```diff
@@ -0,0 +1,119 @@
...
+    // Filter for outgoing stanzas.
+    private static final StanzaFilter OUTGOING_FILTER = new AndFilter(
+            MessageTypeFilter.NORMAL_OR_CHAT_OR_HEADLINE,
+            ToTypeFilter.ENTITY_FULL_OR_BARE_JID);
+
+    // Listener for outgoing stanzas that adds origin-ids to outgoing stanzas.
+    private final StanzaListener stanzaListener = new StanzaListener() {
+        @Override
+        public void processStanza(Stanza packet) {
+            if (!(packet instanceof Message)) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: You don't need this check, it's already done by the filter.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 19]
```diff
@@ -0,0 +1,119 @@
...
+                return;
+            }
+            Message message = (Message) packet;
+            if (!OriginIdElement.hasOriginId(message)) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: Use (and maybe create) a WithOriginIdFilter, and use that filter in combination with the 'not' filter in the OUTGOING_FILTER, then the whole listener comes down to `OriginIdelement.addOriginId(message)`.

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mam/MamIntegrationTest.java [hunk 20]
```diff
@@ -25,7 +25,6 @@
 import org.jivesoftware.smack.SmackException.NotLoggedInException;
 import org.jivesoftware.smack.XMPPException.XMPPErrorException;
 import org.jivesoftware.smack.packet.Message;
-
```
Comment (Flowdalic) [inline smack-integration-test/src/main/java/org/jivesoftware/smackx/mam/MamIntegrationTest.java]: Please re-add the empty line here.

File: smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java [hunk 21]
```diff
@@ -0,0 +1,119 @@
...
+
+    // Filter for outgoing stanzas.
+    private static final StanzaFilter OUTGOING_FILTER = new AndFilter(
+            MessageTypeFilter.NORMAL_OR_CHAT_OR_HEADLINE,
+            ToTypeFilter.ENTITY_FULL_OR_BARE_JID);
+
+    // Listener for outgoing stanzas that adds origin-ids to outgoing stanzas.
+    private final StanzaListener stanzaListener = new StanzaListener() {
+        @Override
+        public void processStanza(Stanza packet) {
```
Comment (Flowdalic) [inline smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: nit: the name of the argument should be 'stanza', not 'packet'.

Comment (vanitasvitae) [reply smack-experimental/src/main/java/org/jivesoftware/smackx/sid/StableUniqueStanzaIdManager.java]: This should also be fixed in the StanzaListener interface.

File: smack-integration-test/src/main/java/org/jivesoftware/smackx/mam/MamIntegrationTest.java [hunk 22]
```diff
@@ -54,7 +54,7 @@ public MamIntegrationTest(SmackIntegrationTestEnvironment environment) throws No
 
     @SmackIntegrationTest
     public void mamTest() throws NotConnectedException, InterruptedException, NoResponseException, XMPPErrorException,
-                    NotLoggedInException {
+            NotLoggedInException {
```
Comment (Flowdalic) [inline smack-integration-test/src/main/java/org/jivesoftware/smackx/mam/MamIntegrationTest.java]: Please revert that whitespace only change

**Comment by Flowdalic** [comment]
Squashed as  b3b76b9ff49b79d4d49ddbb046263c2476f2f3d8

Thanks!
